### PR TITLE
Altair Committee Cache Changes

### PR DESCRIPTION
**Pull request by rauljordan**
This PR adds changes from `beacon-chain/cache` in the `hf1` branch into `develop`, which are necessary for various helpers for the Altair hard fork. This adds committee cache active balance changes

### CODE REVIEW

File: beacon-chain/cache/sync_subnet_ids.go [hunk 1]
```diff
@@ -37,25 +37,22 @@ func (s *syncSubnetIDs) GetSyncCommitteeSubnets(pubkey []byte, epoch types.Epoch
...
-	// Index 0 was used to store validator's join epoch. We do not
-	// return it to the caller.
-	// Index 1 and beyond were used to store subnets.
 	return idxs[1:], types.Epoch(idxs[0]), ok, duration
 }
 
 // GetAllSubnets retrieves all the non-expired subscribed subnets of all the validators
 // in the cache. This method also takes the epoch as an argument to only retrieve
-// assignments for epochs that have happened.
+// assingments for epochs that have happened.
```
Comment (rauljordan) [inline beacon-chain/cache/sync_subnet_ids.go:55]:
```suggestion
// assignments for epochs that have happened.
```

File: beacon-chain/cache/committee.go [hunk 2]
```diff
@@ -127,6 +127,35 @@ func (c *CommitteeCache) ActiveIndices(seed [32]byte) ([]types.ValidatorIndex, e
...
+	item, ok := obj.(*Committees)
+	if !ok {
+		return 0, ErrNotCommittee
+	}
+	if item == nil {
+		return 0, errors.New("item is nil")
+	}
+
+	// Return `ErrNonCommitteeKey` if active balance field doesnt exist in item.
+	if !item.ActiveBalance.Exist {
```
Comment (terencechain) [inline beacon-chain/cache/committee.go:152]: We should test this block

File: beacon-chain/cache/committee_test.go [hunk 3]
```diff
@@ -88,6 +88,24 @@ func TestCommitteeCache_ActiveCount(t *testing.T) {
...
+	balances := &Balance{
+		Exist: true,
+		Total: uint64(10000),
+	}
+	item := &Committees{Seed: [32]byte{'A'}, ActiveBalance: balances}
+	_, err := cache.ActiveBalance(item.Seed)
+	require.Equal(t, ErrNonCommitteeKey, err)
+
+	require.NoError(t, cache.AddCommitteeShuffledList(item))
+
```
Comment (terencechain) [inline beacon-chain/cache/committee_test.go:103]: We should add a test where 
```go
balances := &Balance{
		Exist: false,
	}
```

`ActiveBalance` should return `ErrNonCommitteeKey`

**Review by terencechain — DISMISSED** [review]

File: beacon-chain/cache/sync_subnet_ids.go [hunk 4]
```diff
@@ -37,19 +37,16 @@ func (s *syncSubnetIDs) GetSyncCommitteeSubnets(pubkey []byte, epoch types.Epoch
...
-	// If no committees are saved, we return nothing.
+	// If no committees are saved, we return
+	// nothing.
 	if len(idxs) <= 1 {
 		return []uint64{}, 0, ok, time.Time{}
 	}
-
-	// Index 0 was used to store validator's join epoch. We do not
-	// return it to the caller.
-	// Index 1 and beyond were used to store subnets.
```
Comment (prestonvanloon) [inline beacon-chain/cache/sync_subnet_ids.go:52]: Why are many of these comments removed?

**Review by prestonvanloon — CHANGES_REQUESTED** [review]
It seems like this PR is undoing some of the feedback from #9350 with regards to commentary.

File: beacon-chain/cache/committee.go [hunk 5]
```diff
@@ -127,6 +127,35 @@ func (c *CommitteeCache) ActiveIndices(seed [32]byte) ([]types.ValidatorIndex, e
...
+	} else {
+		CommitteeCacheMiss.Inc()
+		return 0, ErrNonCommitteeKey
+	}
+
+	item, ok := obj.(*Committees)
+	if !ok {
+		return 0, ErrNotCommittee
+	}
+	if item == nil {
```
Comment (prestonvanloon) [inline beacon-chain/cache/committee.go:147]: Can we check that item.ActiveBalance is not nil? Otherwise it could panic further down

**Comment by rauljordan** [comment]
Ready again @prestonvanloon

**Review by prestonvanloon — APPROVED** [review]

**Pull request merged**